skopt/searchcv.py

```diff
@@ -268,43 +266,17 @@ class BayesSearchCV(BaseSearchCV):
 
     """
 
-    def __init__(self, estimator, search_spaces=None, optimizer_kwargs=None,
+    def __init__(self, estimator, search_spaces, optimizer_kwargs=None,
                  n_iter=50, scoring=None, fit_params=None, n_jobs=1,
                  iid=True, refit=True, cv=None, verbose=0,
                  pre_dispatch='2*n_jobs', random_state=None,
                  error_score='raise', return_train_score=True):
 
-        # set of space name: space dict. Stored as a dict, in order
-        # to make it easy to add or remove search spaces, in case
-        # user wants to use `step` function directly.
-        self.search_spaces_ = {}
-        # make sklearn.base.clone work
         self.search_spaces = search_spaces
-
-        # can be None if user intends to provide spaces later manually
-        if search_spaces is not None:
-            # account for the case when search space is a dict
-            if isinstance(search_spaces, dict):
-                search_spaces = [search_spaces]
-            self.add_spaces(list(range(len(search_spaces))), search_spaces)
-
         self.n_iter = n_iter
         self.random_state = random_state
-
-        if optimizer_kwargs is None:
-            self.optimizer_kwargs = {}
-        else:
-            self.optimizer_kwargs = optimizer_kwargs
-
-        # this dict is used in order to keep track of skopt Optimizer
-        # instances for different search spaces. `str(space)` is used as key
-        # as `space` is a dict, which is unhashable. however, str(space)
-        # is fixed and unique if space is not changed.
-        self.optimizer_ = {}
-        self.cv_results_ = defaultdict(list)
-
-        self.best_index_ = None
-        self.multimetric_ = False
+        self.optimizer_kwargs = optimizer_kwargs
+        self._check_search_space(self.search_spaces)
```

Comment:
What do you think of making this `self.search_spaces = self._check_search_space(search_spaces)` and removing L275?

In real scikit-learn style we should not check the arguments here but instead wait till `fit` is called :-/ Maybe that is actually the best thing to do. http://scikit-learn.org/stable/developers/contributing.html#instantiation WDYT?

Reply:
I guess it's OK to run code in the `__init__` that does not alter attributes, a similar check is performed in `GridSearchCV` https://github.com/scikit-learn/scikit-learn/blob/f3320a6f1ab2d060fd49c94a6e3386291a684ec7/sklearn/model_selection/_search.py#L1077